README.md

```diff
@@ -1,5 +1,164 @@
+data.
+
+The shim is divided into two pieces:
+
+* A *binary* that runs in a cross origin iframe, such as
+  `https://fledge-shim.example/v/1234`.
+
+* A *library* that consumers compile into their code, and which communicates
+  with the binary over `postMessage`.
+
+Almost all of the work happens in the binary; the library is a small adapter
+that translate the API from functions to messages.
+
+## API
+
+We're planning to implement the API as closely as possible to what is presented
+in the spec, but some aspects necessarily differ due to the constraints of
+running on publisher and advertiser pages, and implementing in pure JS.
+
+### On-Page API
+
+To make the API available on a page consumers will compile the library into
+their code.  We're thinking of making it available as an NPM package, or people
+can pull from github manually.
+
+#### Initialization
+
+The library needs to know how to load the binary on the page, and this needs to
+happen before any library calls that communicate with a binary.
+
+```javascript
+fledgeShim.initialize('https://fledge-polyfill.example');
+```
+
+The initialization call takes the origin of the binary only. Versioning
+parameters are added automatically.
+
+#### Joining Interest Groups
+
+```javascript
+fledgeShim.joinAdInterestGroup(myGroup, 30 * kSecsPerDay);
+```
+
+Initially, only `owner`, `name`, `user_bidding_signals`, and `ads` will be
+supported.  We also plan to support for `daily_update_url` and
+`trusted_bidding_signals_*`, and at least the `report_win` portion of
+`bidding_logic_url`.  See the worklet discussion below for more on
+`bidding_logic_url`.
+
+The `daily_update_url` is supported, but because we cannot actually run updates
+in the background as the browser would, we will only be able to fetch updates on
+days when the polyfill loads on some page.
+
+For trusted bidding signals, we'll need to implement our own caching of
+responses, since we read FLEDGE as caching at the per-key level.  We'll need to
+parse caching response headers, but we might limit the range of formats we
+accept.
+
+#### Initiating an On-Device Auction
+
+```javascript
+const auctionResultPromise = fledgeShim.runAdAuction(myAuctionConfig);
+```
+
+All auction configuration options will be supported, but `decision_logic_url`
+(see worklet discussion below) will initially only handle `report_result`.
+
+#### Rendering a winning ad
+
+```javascript
+const auctionResultPromise = fledgeShim.runAdAuction(myAuctionConfig);
+auctionResultPromise.then((auctionWinnerUrl) => {
+  // auctionWinnerUrl is an opaque token, and not the real rendering url
+  var adFrame = document.createElement('iframe');
+  // set whatever attributes you like on adFrame, and then call:
+  fledgeShim.renderAd(adFrame, auctionWinnerUrl);
+});
+```
+
+Because fencedframes don't exist yet, this will render the ad in an
+ordinary iframe.  The shim will not be realistic in testing the
```

Comment:
SF?

Reply:
A safeframe is a kind of ordinary iframe